pkg/cluster/service_test.go

```diff
@@ -161,3 +165,128 @@ func TestListToolchainClusterConfigs(t *testing.T) {
 		require.Empty(t, clusterConfigs)
 	})
 }
+
+func TestNewClusterConfig(t *testing.T) {
+	legacyTc := func() *toolchainv1alpha1.ToolchainCluster {
+		return &toolchainv1alpha1.ToolchainCluster{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "tc",
+				Namespace: "ns",
+				Labels: map[string]string{
+					"namespace": "operatorns",
+				},
+			},
+			Spec: toolchainv1alpha1.ToolchainClusterSpec{
+				APIEndpoint: "https://over.the.rainbow",
+				SecretRef: toolchainv1alpha1.LocalSecretReference{
+					Name: "secret",
+				},
+			},
+		}
+	}
+	newFormTc := func() *toolchainv1alpha1.ToolchainCluster {
+		return &toolchainv1alpha1.ToolchainCluster{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "tc",
+				Namespace: "ns",
+			},
+			Spec: toolchainv1alpha1.ToolchainClusterSpec{
+				SecretRef: toolchainv1alpha1.LocalSecretReference{
+					Name: "secret",
+				},
+			},
+		}
+	}
+
+	legacySecret := func() *corev1.Secret {
+		return &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "secret",
+				Namespace: "ns",
+			},
+			Data: map[string][]byte{
+				"token": []byte("token"),
+			},
+		}
+	}
+	kubeconfigSecret := func(t *testing.T) *corev1.Secret {
+		t.Helper()
+		kubeconfig := clientcmdapi.Config{
+			Clusters: map[string]*clientcmdapi.Cluster{
+				"cluster": {
+					Server: "https://over.the.rainbow",
+				},
+			},
+			Contexts: map[string]*clientcmdapi.Context{
+				"ctx": {
+					Cluster:   "cluster",
+					AuthInfo:  "auth",
+					Namespace: "operatorns",
+				},
+			},
+			AuthInfos: map[string]*clientcmdapi.AuthInfo{
+				"auth": {
+					Token: "token",
+				},
+			},
+			CurrentContext: "ctx",
+		}
+		kubeconfigContents, err := clientcmd.Write(kubeconfig)
+		require.NoError(t, err)
+
+		return &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "secret",
+				Namespace: "ns",
+			},
+			Data: map[string][]byte{
+				"kubeconfig": kubeconfigContents,
+			},
+		}
+	}
+
+	t.Run("using legacy fields in ToolchainCluster and token in secret", func(t *testing.T) {
+		tc := legacyTc()
+		secret := legacySecret()
+
+		cl := test.NewFakeClient(t, tc, secret)
+
+		cfg, err := cluster.NewClusterConfig(cl, tc, 1*time.Second)
+		require.NoError(t, err)
+
+		assert.Equal(t, "https://over.the.rainbow", cfg.APIEndpoint)
+		assert.Equal(t, "operatorns", cfg.OperatorNamespace)
+		assert.Equal(t, "token", cfg.RestConfig.BearerToken)
+	})
+
+	t.Run("using kubeconfig in secret", func(t *testing.T) {
+		tc := newFormTc()
+		secret := kubeconfigSecret(t)
+
+		cl := test.NewFakeClient(t, tc, secret)
+
+		cfg, err := cluster.NewClusterConfig(cl, tc, 1*time.Second)
+		require.NoError(t, err)
+
+		assert.Equal(t, "https://over.the.rainbow", cfg.APIEndpoint)
+		assert.Equal(t, "operatorns", cfg.OperatorNamespace)
+		assert.Equal(t, "token", cfg.RestConfig.BearerToken)
+	})
+
+	t.Run("uses kubeconfig in precedence over legacy fields", func(t *testing.T) {
+		tc := newFormTc()
```

Comment:
do we need to add a test case that uses the `legacyTc`? 

basically:
```
                 tc := legacyTc()
		// Combine the kubeconfig and the token in the same secret.
		// We should see auth from the kubeconfig used...
		secret := kubeconfigSecret(t)
		secret.Data["token"] = []byte("not-the-token-we-want")
```

Reply:
This essentially checks that if we detect a `kubeconfig` data field in the secret then we use that instead of the data in the `ToolchainCluster` and the token in the secret.

So as soon as we see the data coming from the kubeconfig, we know that happened.

Now, you have a very good idea for a "defensive test" that would make sure any future changes wouldn't change that behavior. But we plan to completely remove the loading from the legacy fields in the upcoming PRs so I think we can afford to not have that and instead just make sure we use the new over the old.

Reply:
Right, I was asking since as soon as you push this PR to prod, the tc will still have the `legacy` format, and later will have the new one. But I guess it's not a concern since the only thing that matters is the loading mechanism and not the tc format.